Added -fcas-emit-casid-file to emit a .casid file

rastogishubham opened the pull request:
With -fcas-emit-casid-file, a .casid file will be emitted next to the object file only when -fcas-backend-mode=native, or -fcas-backend-mode=verify is used and the output is directed to a file, rather than stdout.

rastogishubham commented:
@swift-ci please test llvm

adrian-prantl commented:
Nice! It might make future cherry-picking easier if we separate the clang and llvm changes into separate commits.

cachemeifyoucan commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -1145,7 +1149,7 @@ void EmitAssemblyHelper::RunCodegenPipeline(
         return;
     }
-    if (!AddEmitPasses(CodeGenPasses, Action, *OS,
+    if (!AddEmitPasses(CodeGenPasses, Action, *OS, &(*CasIDOS),
```
nit: CasIDOS.get()

cachemeifyoucan reviewed (commented):
LGTM in general.

Other suggestion I will make is to make all the CASIDOS to be the last parameter with a default value (nullptr) when possible. That will keep the diff a lot smaller downstream. It can be re-ordered when upstreaming.

Also since this is a LLVM CodeGen option, I prefer the bulk of the testing should be done in llvm test. That is using `llvm-mc` or `llc` to test CASID file output. Instead of using clang to do all the testing.

rastogishubham committed:
Add -fcas-emit-casid-file to emit a .casid file

With -fcas-emit-casid-file, a .casid file will be emitted next to the
object file only when -fcas-backend-mode=native, or
-fcas-backend-mode=verify is used and the output is directed to a file,
rather than stdout.

rastogishubham committed:
Enable -fcas-emit-casid-file to llvm backend.

This change enables -fcas-emit-casid-file to the llvm backend and allows
emission of a .casid file when -fcas-backend and
-fcas-backend-mode=native or -fcas-backend-mode=verify is used

rastogishubham commented:
@swift-ci please test

rastogishubham commented:
@swift-ci please test llvm

rastogishubham commented:
@swift-ci please test llvm
